docs/module-caching.md

```diff
@@ -0,0 +1,181 @@
+   const { check: c2 } = await import(mod);
+   assert(c1 === c2);
+
+   // dep.js
+   export const mod = module { export const check = {}; };
+   export const { check } = await import(mod);
+   ```
+4. A module block is evaluated in the same realm as where it is imported:
+   ```js
+   const mod = module { globalThis.modEvaluated = true; };
+   const realm = createLegacyRealm(); // [1]
+   await realm.eval(`s => import(s)`)(mod);
+   assert(globalThis.modEvaluated === undefined);
+   assert(realm.modEvaluated === true);
+   ```
+   This is consistent with the behavior of imports with string specifiers.
+5. Importing the same module block twice _from different Realms_ results in multiple evaluations:
+   ```js
+   const realm = createLegacyRealm(); // [1]
+   globalThis.count = 0;
+   realm.count = 0;
+   const mod = module { globalThis.count++; export const check = {}; };
+   const { check: c1 } = await import(mod);
+   const { check: c2 } = await realm.eval("s => import(s)")(mod);
+   assert(globalThis.count + realm.count === 2);
+   assert(c1 !== c2);
+   ```
+   This is consistent with the behavior of imports with string specifiers.
+6. Importing two different module blocks results in two evaluations and two different namespace objects. Two module blocks `mod1` and `mod2` are considered different if `mod1 !== mod2`:
+   ```js
+   globalThis.count = 0;
+   const mod1 = module { globalThis.count++; export const check = {}; };
+   const mod2 = module { globalThis.count++; export const check = {}; };
+   const { check: c1 } = await import(mod1);
+   const { check: c1 } = await import(mod2);
+   assert(c1 !== c2);
+   assert(globalThis.count === 2);
+   ```
+   ```js
+   globalThis.count = 0;
+   const mod = module { globalThis.count++; export const check = {} };
+   const modClone = structuredClone(mod);
+   assert(mod !== modClone);
+   const { check: c1 } = await import(mod);
+   const { check: c2 } = await import(modClone);
+   assert(c1 !== c2);
+   assert(globalThis.count === 2);
+   ```
+   ```js
+   globalThis.count = 0;
+   const mod = module {};
+   const worker = new Worker(`data:text/javascript,
+     addEventListener("message", msg => postMessage(msg.data))
+   `);
+   worker.addEventListener("message", ({ data: mod2 }) => {
+     assert(mod !== mod2);
+     const { check: c1 } = await import(mod);
+     const { check: c2 } = await import(mod2);
+     assert(c1 !== c2);
+     assert(globalThis.count === 2);
+   });
+   worker.postMessage(mod);
+   ```
+
+## Invariants enforced by the HTML specification
+
+These invariants cannot be enforced by ecma262, since it doesn't define how cloning works and how string specifiers are resolved. They will be respected by the HTML integration, and the champion group suggests that hosts that have a similar modules model could follow these patterns.
+
+7. When serializing&deserializing a module block, the "referrer" used as the base to resolve string specifiers should be kept the same:
```

Comment:
```suggestion
7. When serializing and deserializing a module block, the "referrer" used as the base to resolve string specifiers should be kept the same:
```

Reply:
I think we can specify this in ECMA-262 and without needing to specify serialization or string specifier resolution.

Basically, if we can relate the concept of a `"compiled module record"` as existing across realms, then a module block can be associated against its parent source text compiled module record both of which can be fully defined ECMA-262 constructs. The resolution idempotence is then that:

> If compiled module record `b` is a module block, which is defined within a source text module compiled module record `s`, then for any realm `r` and specifier `specifier`, `HostResolveImportedModule(specifier, b, r) === HostResolveImportedModule(specifier, s, r)` where `HostResolveImportedModule` is updated to itself return a compiled module record which is then singly instanced within the realm. This is a fully-well defined invariant definition in ECMA-262. Separately, specs like compartments can then tackle the problem of multiple instancing of a compiled module within the same realm and how to relax the constraint again.

Reply:
The current thinking, between @caridy and myself, is that we will bring back you `ModuleInstance` sketch, such that each module instance has its own memoization for its `importHook`. We would introduce a *_Virtual Module Record*_ that manages the lifecycle and contains a reference to a *_Static Module Record_* which, as you describe, would be immutable and safely shared between realms in the same process. *_Source Text Static Module Record_* would initially be the only concrete implementation.

That is to say, we would not alter per-realm behavior at all and that we would enforce invariance on a per-instance basis, recovering sufficient idempotence in aggregate by induction over the import hook.

Reply:
It makes sense that for arbitrary resolution hooks an instance model is the idempotence model.

But it's important to note that these are two separate resolution models - one is the host default resolution model, and one is a custom resolution model.

One nice thing about maintaining the concept of a singular compartment, is that the compartment itself forms the unit of idempotence, whereas with a full instance model that boundary is no longer clear.

Therefore different constraints might apply at the uninstrumented host level, the compartment level, or the arbitrary instance level (from most constrainted idempotence to the least).

Reply:
Regardless how the above works out, I think it makes sense for the host contraints for the default resolution model, and an instrumented resolution model to be clearly different resolution constraint scenarios.

Reply:
> Basically, if we can relate the concept of a "compiled module record" as existing across realms, then a module block can be associated against its parent source text compiled module record both of which can be fully defined ECMA-262 constructs. The resolution idempotence is then that:

How much "global" can something be? Can something also exist across agents? For example, web workers are different agents from the main one.

Reply:
Actually, I think that the proposed constraint is ok as-is and doesn't need to be relaxed for compartments/loaders. `structuredClone` will keep the same referrer, but every compartment can map a (_referrer_, _specifier_) to something different. This is not visible in this proposal, because currently there is 1 compartment per Realm: the cache map will need to be moved from the Realm to the Compartment.